Bluetooth: Mesh: Refactor proxy adv

Andrewpini opened the pull request:
Refactors proxy advertising implementation to allow fair sharing of advertising resources between all subnets.

In the new implementation, each subnet is agnostic to any other subnet that might have active proxy advertisement work. When proxy advertisement is triggered, the implementation will first check how many subnets that has active work. If there are more than one active subnet, a maximum timeslot will be calculated to ensure that all active subnets get access to the medium. The implementation will then poll the next eligible subnet for a proxy advertising request. If the duration of this request exceeds the maximum timeslot, the duration for the next advertisement will be set to the maximum value.

If a proxy advertisement for a subnet is interrupted by other advertising activity, the implementation will now ensure that the interrupted proxy adv continues from the point where it was interrupted so that the subnet gets to utilize the entire allocated timeslot.

This PR also alters the priv_proxy_net_id_multi Bsim test to align with the refactored proxy advertising scheme, and adds test that verifies correct Proxy advertisement behavior for a device where the Proxy adv requirements changes over time, both for single and multiple subnets.

alxelax commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+
+	int32_t max_adv_duration;
+	int active_sub_cnt;
+	static int32_t prev_start;
```
There are 4 static variables those characterize the state of the previous subnetwork advertising.
I recommend creating something like:
```
static struct {
    int32_t prev_start;
    struct bt_mesh_subnet *prev_sub;
    struct proxy_adv_req prev_req;
    bool ignore_previous;
} prev_sub_adv;
```
and keep it outside the function. It will help to keep this information in one place and make it available from for debugging\logging in other places.

alxelax reviewed (changes requested):
At the first glance it seems Ok. I do not have any objections to make proxy advertising more structured since it has quite messed implementation. 
However, it has so poor test coverage, so it is very difficult to say it breaks anything or doesn't. Despite the mess of implementation, the previous version worked somehow. 
These changes require more tests, obviously.

PavelVPV commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+				sub = adv_sub_get(prev_sub, &request, true);
 
-		if (sub->node_id == BT_MESH_NODE_IDENTITY_RUNNING) {
-			uint32_t active = k_uptime_get_32() - sub->node_id_start;
-			bool priv_node_id = false;
+				/* Check that it is the same event */
+				if ((sub == prev_sub) && (prev_req.evt == request.evt)) {
```
Can't this call with `specific == true` be simply replaced with `adv_sub_check`? Because, unless I'm missing something, `while (sub_temp != sub_start && !specific)` exits immediately if `specific == true`. Then `adv_sub_get` either returns `prev_sub` or `NULL`.

Andrewpini commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+		 */
+		if (req->duration < MSEC_PER_SEC) {
+			bt_mesh_proxy_identity_stop(sub);
+			LOG_DBG("Node ID stopped for subnet %d", sub->net_idx);
```
I'll change it to something other than `_check` 👍

Andrewpini commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
-		}
-	}
+		/* Check if the previous subnet finished its allocated timeslot */
+		if (is_sub_proxy_active(prev_sub) && prev_req.duration != SYS_FOREVER_MS) {
```
Im not sure I got your meaning here.

PavelVPV replied:
Ah, ok, I was confused by comment at line 806. It stops advertising for those subnets that advertise for more than `NODE_ID_TIMEOUT`. If `prev_sub` advertises for less than `NODE_ID_TIMEOUT`, `is_sub_proxy_active` will return true. Then the code is OK.

Andrewpini commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
-		}
-	}
+		/* Check if the previous subnet finished its allocated timeslot */
+		if (is_sub_proxy_active(prev_sub) && prev_req.duration != SYS_FOREVER_MS) {
```
Checked. It will.

Andrewpini commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+	default:
+		LOG_ERR("Unexpected proxy adv evt: %d", request.evt);
+		return -ENODEV;
+	}
```
I will look into it 👍

Andrewpini commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+		 *  prevent this we reset the Node Identity adv state
+		 *  here, but let it advertise for one additional second.
+		 */
+		if (req->duration < MSEC_PER_SEC) {
```
I will look into it 👍

Andrewpini committed:
Bluetooth: Mesh: Refactor proxy adv

Refactors proxy advertising implementation to allow fair sharing of
advertising resources between all subnets.

In the new implementation, each subnet is agnostic to any other subnet
that might have active proxy advertisement work. When proxy
advertisement is triggered, the implementation will first check how
many subnets that has active work. If there are more than one active
subnet, a maximum timeslot will be calculated to ensure that all active
subnets get access to the medium. The implementation will then poll
the next eligible subnet for a proxy advertising request. If the
duration of this request exceeds the maximum timeslot, the duration
for the next advertisement will be set to the maximum value.

If a proxy advertisement for a subnet is interrupted by other
advertising activity, the implementation will now ensure that the
interrupted proxy adv continues from the point where it was
interrupted so that the subnet gets to utilize the entire allocated
timeslot.

This PR also alters the priv_proxy_net_id_multi Bsim test to align
with the refactored proxy advertising scheme.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini committed:
tests: Bluetooth: Mesh: Add proxy adv coex test.

Adds test that verifies correct Proxy advertisement behavior for
a device where the Proxy adv requirements changes over time,
both for single and multiple subnets.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,250 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+				request.duration = prev_sub_adv.req.duration - time_delta;
+				if (request.duration >= MSEC_PER_SEC) {
+					goto end;
+				} else if (!prev_sub_adv.ignore_previous) {
```
This is because adv interval is 1 second, so if timeout is less than interval, the controller ends the advertisement immediately, right? What if we add some margin to the timeout, e.g. 10ms?

Andrewpini replied:
>Can't proxy_adv_sub_request_get simply return false here

`proxy_adv_sub_request_get` returns the requested time for advertising proxy beacons from a single subnet, which is not necessarily the adv time it will be given if there are multiple subnets with proxy adv work.

PavelVPV replied:
10ms is minimal because the timeout value is in 10ms step:
https://github.com/zephyrproject-rtos/zephyr/blob/ed610400bcb74338d25e4b29f948709efe12f3a3/subsys/bluetooth/mesh/adv_ext.c#L547

Andrewpini commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,250 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
-	sub = beacon_sub ? beacon_sub : bt_mesh_subnet_next(beacon_sub);
-	if (!sub) {
-		LOG_WRN("No subnets to advertise on");
+	active_sub_cnt = active_proxy_sub_cnt();
```
Well, I see that in this case it is a little more messy than usual. I will try to clean up and be a bit more consistent.

alxelax commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +658,250 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
+		if ((prev_sub_adv.req.duration != SYS_FOREVER_MS) &&
+		    proxy_adv_sub_request_get(prev_sub_adv.sub, &request) &&
+		    (prev_sub_adv.req.evt == request.evt)) {
+			int32_t time_delta = k_uptime_get_32() - prev_sub_adv.start;
```
elapsed

alxelax commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -626,209 +626,255 @@ static int net_id_adv(struct bt_mesh_subnet *sub, int32_t duration)
+		if (time_passed > ((MSEC_PER_SEC * bt_mesh_od_priv_proxy_get()) - MSEC_PER_SEC)) {
+			sub->priv_net_id_sent = 0;
+			sub->solicited = false;
+			LOG_DBG("Advertising Private Network ID timed out after solicitation");
```
I guess it is worth adding elapsed time and subnetwork index in logging too.

alxelax commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -626,209 +626,255 @@ static int net_id_adv(struct bt_mesh_subnet *sub, int32_t duration)
-	if (subnet_count > 1) {
-		int32_t max_timeout;
+	} else if (cnt > 1) {
+		/** There are more than one subnet that requires proxy adv,
```
```suggestion
		/** There is more than one subnet that requires proxy adv,
```

alxelax commented on tests/bsim/bluetooth/mesh/src/test_beacon.c:
```diff
@@ -1633,15 +1649,67 @@ static void priv_scan_cb(const bt_addr_le_t *addr, int8_t rssi, uint8_t adv_type
+struct proxy_adv_beacon {
+	uint8_t evt_type;
+	uint8_t net_idx;
+	int64_t recv_ts;
```
```suggestion
	int64_t rx_timestamp;
```

alxelax commented on tests/bsim/bluetooth/mesh/src/test_beacon.c:
```diff
@@ -1781,6 +1843,211 @@ static void test_tx_priv_multi_net_id(void)
+		proxy_adv_subnet_find(&beac, nets, net_cnt);
+		proxy_adv_register_evt(&beac, exp_evt, exp_evt_cnt);
+
+		/* Sleep a little less than one second */
```
this is obviously from the code. :) The most interesting knowledge why?

alxelax reviewed (commented):
LGTM just minor comments

Andrewpini commented on tests/bsim/bluetooth/mesh/tests_scripts/priv_beacon/proxy_adv_multi_subnet_coex.sh:
```diff
@@ -0,0 +1,55 @@
+# the RX device scans and verifies that the correct proxy adv messages of
+# the different subnets is sent within the expected time delta.
+
+# Note 1: The maximum allowed timeslot for a subnet to
+# advertise proxy in this scenario is 10 seconds when there is more than
```
"The maximum allowed timeslot for a subnet to advertise proxy in this scenario is 10 seconds **when there is more than one subnet that has active proxy adv work.**"

I thought that it was already stating that it applied to multi subnet scenario, but if it is unclear I can rephrase it. The main purpose of this note is to emphazies that in this particular test scenario the max TS is 10 seconds, so that the reader might get the timing/adv_count criteria in the test. For a test/usecase with more than 6 subnets, with active proxy adv, the max TS wil be less than 10 seconds.

BTW, do you know if there is some reasoning behind the `max_adv_duration = NODE_ID_TIMEOUT / MAX(cnt, 6);`? It is a remnant from MshPRF1.0.1, so does this calculation fully consider that we now have proxy advertising that depends on the On-Demand Private GATT Proxy state (1-255 seconds)?

PavelVPV replied:
> I thought that it was already stating that it applied to multi subnet scenario, but if it is unclear I can rephrase it. The main purpose of this note is to emphazies that in this particular test scenario the max TS is 10 seconds, so that the reader might get the timing/adv_count criteria in the test. For a test/usecase with more than 6 subnets, with active proxy adv, the max TS wil be less than 10 seconds.

Ah, ok, true, my mistake.

> BTW, do you know if there is some reasoning behind the max_adv_duration = NODE_ID_TIMEOUT / MAX(cnt, 6);? It is a remnant from MshPRF1.0.1, so does this calculation fully consider that we now have proxy advertising that depends on the On-Demand Private GATT Proxy state (1-255 seconds)? As an example, a device has two subnets, where the first subnet has ongoing active proxy adv work, and then Private Network Identity adv is triggered on the second subnet via solicitation. With the suggested implementation they will get a 10 sec timeslot each, meaning that if the On-Demand Private GATT Proxy state is set to lets say 17 seconds it will in reality only advertise for 10 seconds.

I don't know why it is split into at least 6 slices. Perhaps to alternate between subnets more often, otherwise, e.g. with 2 subnets, you need to wait for 30 seconds before seeing another subnet advertisements. Regarding the solicitation, as we discussed offline, the solicited subnet should get exclusive rights to the advertiser. Perhaps, the implementation should consider that the solicited subnet adv duration can be more than 60 seconds.

> Is there a good reason why the proxy adv internals should not be triggered more frequently, lets say with a 2 second interval. Of course it will cause some extra power and resource consumption, but is it substantial? Should we consider a Kconfig option for the max duration rather than a fixed one?

I think a GATT client may have less chance to connect to the device.

PavelVPV commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -626,209 +626,255 @@ static int net_id_adv(struct bt_mesh_subnet *sub, int32_t duration)
-		if (remaining > max_timeout || remaining == SYS_FOREVER_MS) {
-			remaining = max_timeout;
+		max_adv_duration = NODE_ID_TIMEOUT / MAX(cnt, 6);
+		max_adv_duration = MAX(max_adv_duration, MSEC_PER_SEC);
```
So it worked?

Andrewpini replied:
I'm working on running the Bsim with SDC

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_beacon.c:
```diff
@@ -1542,60 +1543,78 @@ static void test_tx_priv_beacon_cache(void)
 
+static uint8_t test_net_key_3[16] = {0x12, 0x54, 0xab, 0x1e};
+
+#define UNTIL_UPTIME(time) (k_uptime_get() > time ? K_NO_WAIT : K_MSEC(time - k_uptime_get()))
```
```suggestion
#define UNTIL_UPTIME(time) (k_uptime_get() > (time) ? K_NO_WAIT : K_MSEC((time) - k_uptime_get()))
```

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_beacon.c:
```diff
@@ -1633,15 +1652,61 @@ static void priv_scan_cb(const bt_addr_le_t *addr, int8_t rssi, uint8_t adv_type
+		return;
+	}
+
+	beac->evt_type = proxy_adv_type_pull(buf);
+	if (beac->evt_type > BEACON_TYPE_PRIVATE_NODE_ID) {
+		return;
+	}
```
`proxy_adv_type_check` could be changed to return beacon type.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_beacon.c:
```diff
@@ -1781,6 +1840,246 @@ static void test_tx_priv_multi_net_id(void)
+		 .time = {.after = PROXY_ADV_MULTI_CHECKPOINT_2,
+			  .before = PROXY_ADV_MULTI_CHECKPOINT_3}},
+
+		/** The first and second subnet gets solicited. Check that
+		 *  PRIVATE_NET_ID is advertised by these subnet,
+		 */
```
This is again against this paragraph:
> When the server starts advertising with Private Network Identity as a result of the reception of a Solicitation PDU (see Section 6.9.1), the server **shall only** advertise the subnet corresponding to the network key that is used to authenticate the Solicitation PDU; otherwise, the server shall interleave the advertising of each of its subnets. The duration of the advertising is not limited.

The word `only` makes me feel that only one subnet can be advertised at a time and the last solicited subnet "cancels" previous solicitation request. @omkar3141 , can you clarify this?

PavelVPV replied:
And this phrase I couldn't find, but now I found it (section 7.2.2.2.1):
> A node may advertise with Node Identity or Private Node Identity regardless of the Proxy feature being supported or enabled

Which means either node id or network id, not both

omkar3141 replied:
About "only":

The phrasing "shall only advertise" here means: Advertise for all of those for which solicitation PDU is received. This phrasing is used at multiple places in the spec to emphasize that, one-to-one relationship between the netID advs for a given subnet and solicited subnet.

You can have following situation that should work properly for network users:

* Assume the following: A node has multiple NetKeys, users of the network are given seperate network keys depending on their access levels. 
* Now, if all users gather at same place and want to connect via same proxy (provided sufficient connections number of connections are supported on that node), they would send solicitation PDU and should be able to connect. Not advertising PrivNetID for each solicited subnet would break the behavior for some users.

PavelVPV replied:
Ah, sorry, this sentence has different meaning. You can ignore that comment. The question about advertising (private) node id and private network id at the same time is still open...

PavelVPV replied:
Discussed offline with @omkar3141. Regarding "only" word in that paragraphs: @omkar3141 will file an erratum. For now lets leave this as is and allow to solicit multiple subnets at the same time. Regarding interleaving net id and node id: lets allow it as well. They are 2 different features and the spec doesn't states clearly that this is not allowed.

PavelVPV commented on subsys/bluetooth/mesh/proxy_srv.c:
```diff
@@ -680,155 +654,225 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
-	sub = beacon_sub ? beacon_sub : bt_mesh_subnet_next(beacon_sub);
-	if (!sub) {
-		LOG_WRN("No subnets to advertise on");
+	cnt = active_proxy_sub_cnt_get();
```
Please disregard this: https://github.com/zephyrproject-rtos/zephyr/pull/65914/files#r1426738862
